### ISSUE TITLE
Unit Tests

### ISSUE THREAD

**Issue by tidyui** [PiranhaCMS/piranha.core#15]
Unit tests for all core functions as well as for the data services.

**Comment by TheWolfA2** [comment]
Will this follow the .Net Core "standard" of a test folder as a sibling to the src folder? Similarly, tests I presume will follow the convention Piranha.<Module>.Tests, correct?

**Comment by tidyui** [comment]
Yes exactly! As long as the unit tests are cross-platform, I've only use the built in test projects in Visual Studio in the past which don't load on Mac for example.

**Comment by TheWolfA2** [comment]
xunit does wonders, I can set that up so that there is at least the core framework and setup for unit testing is in place

**Comment by TheWolfA2** [comment]
As I'm working on unit tests I've come across a question I have about some of the manager area controllers. So all of the controllers posses an `IApi` instance which I would presume is for DI purposes. However, inside of the various action methods, say `BlockTypeController.List`, the model data is obtained through `Piranha.App` without a call to `Piranha.App.Init`, making the `IApi` injection useless as far as I can tell. What is the reason for this setup?

On a separate note, I've also updated the manager area controllers to derive from an abstract base controller so that these common properties and attributes are easily updated from one file instead of across multiple controller files. Hopefully this design choice doesn't cause a major issue?

**Comment by tidyui** [comment]
The `Piranha.App` object is a singleton which is initialized once during the application lifetime cycle. This is done in the `Startup.cs` of the current application in the `Configure` method, for example:

https://github.com/PiranhaCMS/piranha.core/blob/master/src/examples/Blog/Startup.cs#L54

I'm guessing the call to `App.Init()` should be done in the unit test setup before the tests start just like it's done in the web app startup.

The PageTypes & BlockTypes are cached in the application and loaded from the repositories during initialization, but it can always be discussed if the manager should load them from cache or get them directly from data storage. However in the first verision, both the manager views for **BlockTypes** and **PageTypes** will most likely be skipped as they are currently only created with JSON or Attributes.

Regards

**Comment by tidyui** [comment]
And a base class for the manager controller sounds good, no issues there!

**Comment by TheWolfA2** [comment]
Ahh! Okay, that's the link I was missing! That makes more sense now. Yeah, I think `App.Init()` needs to be called within the test setup then.

My only concern with caching in `App` is it creates an inconsistent pattern of how data in the manager controllers are accessed. I would argue it's better to keep it consistent across the board. However, if **PageType** and **BlockType** are just special cases and it's well documented that would be sufficient I think

**Comment by TheWolfA2** [comment]
I'm encountering an issue now with calling `App.Init()`, specifically the values being cached from the `IApi` object are persisting between tests. For example, `BlockTypeControllerUnitTest` should have 0 `PageType` objects and 5 `BlockType` objects, and `PageControllerUnitTest` should have 5 `PageType` objects and 0 `BlockType` objects. There could be a common set of data created for all unit tests, however unit tests shouldn't rely on data between one another.

How about an `internal static void ForceInit()` that is exposed to the unit tests which ignores the `App.isInitialized`, and recreates the `App.instance` object? Then that can be called instead of `App.Init` for each test initialization.

**Comment by tidyui** [comment]
Maybe the unit test could use the BlockType/PageType repositories instead of using the cached objects in the App singleton?

Also once the first level caching is in place we could consider moving them out of the App object and instead cache them within the repository.

Regards

**Comment by TheWolfA2** [comment]
I'm not quite sure I understand what you mean and how this would fix the issue? Are you suggesting the scheme for caching and getting block/page types moves away from the App class?

Realistically, there's no issue with the App singleton itself of its usage in the unit tests. The problem is only that data is persisting between tests, but the singleton needs to behave as though each unit test is a separate instance of Piranha running and therefore would need to call `Init()` every time, but the `!App.isInitialized` check prevents it from reinitializing.

**Comment by tidyui** [comment]
Well the **repositories** have no caching at this point, so all calls made to the `api.PageTypes` and `api.BlockTypes` will get the data straight from the database. There is another issue #9 for creating First level caching, which there's a initial test for in a separate branch. The caching would then take place in the repositories themselves, and if this would be implemented for the PageType & BlockType repositories then the properties in the App object would be redundant.

The current properties in the App object where put there before we took the decision to persist types in the database, in the first version the types were built on startup and only kept in memory.

Regards

**Comment by TheWolfA2** [comment]
Okay, so it sounds like the general caching mechanism is still being defined and developed, but there will be a movement away from the `App` model.

With that in mind, I'll stick to continuing to develop unit tests the way I am that way we have verifiable tests once the caching is complete.

What's good is the unit tests, if run individually, are correct! Running them successively is the problem, and we can address the issue once caching is better defined :)

**Comment by tidyui** [comment]
How's progress @TheWolfA2! If you have some tests ready it's easier to merge a couple at a time, otherwise the pull request will be huge :)

**Comment by TheWolfA2** [comment]
Tests are going well! Most of the infrastructure for Piranha.EF and its repository testing is in place and PageTypeRepository has a solid set of unit tests. Still need to handle the Save method unit tests. I'll send a PR and merge those over today! I also have some Piranha.Manager tests as well for the page type controller, but I'll make a separate pull for that. Piranha.Manager does have some failing tests, which I documented [here](https://github.com/TheWolfA2/piranha.core/issues/3) and [here](https://github.com/TheWolfA2/piranha.core/issues/1) which should be addressed

**Comment by tidyui** [comment]
How do you run the tests @TheWolfA2. I know I can run them from the console when I'm located in the project directory by typing:

    dotnet test

But is there any structured way to run all of the tests and get a structured summary (other than using Visual Studio of course)

Regards

**Comment by TheWolfA2** [comment]
`.vscode/tasks.json` contains a task for each of the test project in the test folder.  If you perform `Run Task` within VSCode you can select one of those projects and it will run all tests within that project. Unfortunately I couldn't figure out how to run all tests at once yet (it may require a `project.json` in the `test/` folder) but running entire projects in an option.

From console you can also run individual classes with the following
```
dotnet test -class Full.Namespace.To.MyClass
```

**Comment by TheWolfA2** [comment]
What should be done about unit tests that are failing?  Should suggestions be noted in a PR the unit tests are coming from, should fixes be implemented in a separate branch and sent in a separate PR, or something else entirely?

So far the only unit tests I have failing are the `Save()`, `Publish()`, and `UnPublish()` methods of the `PageController`.  They are failing for less substantial reasons, but I'm of the opinion that no exceptions or errors should be without a `catch` statement and handling.

**Comment by tidyui** [comment]
Yes make a note of it in the PR. The goal should of course be no failing tests so comments on those that are failing are good. Also, I've added a new issue label here for `tests` so we can add individual issues for different tests.

/Håkan

**Comment by tidyui** [comment]
Current code coverage is good :) Now we just need to make sure we keep it up.
